algebra-core/src/curves/mod.rs

```diff
@@ -308,3 +310,218 @@ where
         Fq = <Self::E1 as PairingEngine>::Fr,
     >;
 }
+
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    // This function consumes the scalars
+    // We can make this more generic in the future to use other than u16.
+
+    // TODO: Generalise to A != 0
+    // Computes [-p, p, -3p, 3p, ..., -2^wp, 2^wp]
+    fn batch_wnaf_tables(bases: &[Self], w: usize) -> Vec<Vec<Self>> {
+        let half_size = 1 << w;
+        let batch_size = bases.len();
+
+        let mut tables = vec![Vec::<Self>::with_capacity(half_size); batch_size];
+
+        let mut a_2 = bases[..].to_vec();
+        let mut tmp = bases[..].to_vec();
+
+        Self::batch_double_in_place(&mut a_2, (0..batch_size).collect());
+
+        for i in 0..half_size {
+            if i != 0 {
+                Self::batch_add_in_place(
+                    &mut tmp,
+                    &mut a_2.to_vec()[..],
+                    (0..batch_size).map(|x| (x, x)).collect(),
+                );
+            }
+
+            for (table, p) in tables.iter_mut().zip(&tmp) {
+                table.push(p.clone());
+            }
+        }
+        tables
+    }
+
+    // This function mutates the scalars in place
+    // We can make this more generic in the future to use other than u16.
+    fn batch_wnaf_opcode_recoding<BigInt: BigInteger + AsRef<[u64]>>(
```

Comment:
`BigInteger` already requires `AsRef<[u64]>`, so you don't need to have another impl

```suggestion
    fn batch_wnaf_opcode_recoding<BigInt: BigInteger>(
```